### PR TITLE
Add AssetSelection.key_prefixes

### LINKED ISSUE #14151: ability to do AssetSelection.key_prefix(..)
### What's the use case?

We generally structure our key prefixes similar to catalogs and schemas on the underlying data structure and by source systems. Rather than having to create a job from a mixture of groups, for instance, it would be neat to have something like 

```python

   ingress_job = define_asset_job(
        name="ingress_finance_from_dw",
        selection=AssetSelection.key_prefix(["old_datawarehouse", "finance"]),
    )
  ```

### Ideas of implementation

_No response_

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Adds an `AssetSelection.key_prefixes` method to match assets based on key prefix. Resolves #14151.

## How I Tested These Changes

New unit test.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #14163** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14163" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/14163?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/asset_selection.py [hunk 1]
```diff
@@ -97,6 +97,24 @@ def keys(*asset_keys: CoercibleToAssetKey) -> "KeysAssetSelection":
         ]
         return KeysAssetSelection(*_asset_keys)
 
+    @public
+    @staticmethod
+    def key_prefixes(*key_prefixes: CoercibleToAssetKey) -> "KeyPrefixesAssetSelection":
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_selection.py:102]: This should be CoercibleToAssetKeyPrefix, right?

File: python_modules/dagster/dagster/_core/definitions/asset_selection.py [hunk 2]
```diff
@@ -97,6 +97,24 @@ def keys(*asset_keys: CoercibleToAssetKey) -> "KeysAssetSelection":
...
+        """Returns a selection that includes assets that match any of the provided key prefixes.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_selection.py:103]: The docstring should somehow indicate that this accepts lists of strings, in addition to stirngs.

**Review by sryza — CHANGES_REQUESTED** [review]
Nice - I think people have been asking for this for a while.

File: python_modules/dagster/dagster/_core/definitions/asset_selection.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/asset_selection.py:102]: huh, didn't know that existed

File: python_modules/dagster/dagster/_core/definitions/asset_selection.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/asset_selection.py:103]: added an example

File: python_modules/dagster/dagster/_core/definitions/events.py [hunk 3]
```diff
@@ -190,18 +193,23 @@ class AssetKeyPartitionKey(NamedTuple):
...
-    else:
+    try:
+        return key_prefix_from_coercible(prefix) if prefix is not None else None
+    except check.CheckError:
         raise check.ParameterCheckError(
             f'Param "{param_name}" is not a string or a sequence of strings'
         )
 
 
+def key_prefix_from_coercible(key_prefix: CoercibleToAssetKeyPrefix) -> Sequence[str]:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/events.py:204]: Nice

**Pull request merged**